engine/apps/schedules/quality.py

```diff
@@ -0,0 +1,322 @@
+    datetime.time object with adjustment according to user.timezone.
+    """
+    if weekday in ["saturday", "sunday"]:
+        # assuming that no working hours are possible for saturday and sunday, this may change when there's a way to
+        # change working hours in the UI
+        return datetime.time(0, 0, 0), datetime.time(0, 0, 0)
+
+    start = user.working_hours[weekday][0]["start"]
+    end = user.working_hours[weekday][0]["end"]
+
+    start_hour, start_minute, start_second = map(int, start.split(":"))
+    end_hour, end_minute, end_second = map(int, end.split(":"))
+
+    tzinfo = pytz.timezone(user.timezone or "UTC")
+
+    start_result = (
+        datetime.datetime.now(tz=tzinfo)
+        .replace(hour=start_hour, minute=start_minute, second=start_second, microsecond=0)
+        .astimezone(pytz.UTC)
+        .time()
+    )
+    end_result = (
+        datetime.datetime.now(tz=tzinfo)
+        .replace(hour=end_hour, minute=end_minute, second=end_second, microsecond=0)
+        .astimezone(pytz.UTC)
+        .time()
+    )
+
+    return start_result, end_result
+
+
+def spans_inside_working_hours(
+    start: datetime.datetime,
+    end: datetime.datetime,
+    user: User,
+) -> list[tuple[datetime.datetime, datetime.datetime]]:
+    """
+    Get spans of events inside working hours for particular event and user.
+    E.g. if working hours are from 9:00 to 17:00, and event is from 7:00 to 10:00,
+    the result will be equal to [(datetime.time(9, 0), datetime.time(10, 0))].
+    """
+
+    days = split_to_days(start, end)
+    result = []
+
+    for day in days:
+        working_hours_start, working_hours_end = get_working_hours(user, day.weekday)
+
+        work_start = datetime.datetime.combine(
+            day.start,
+            working_hours_start,
+            tzinfo=pytz.UTC,
+        )
+
+        work_end = datetime.datetime.combine(
+            day.end,
+            working_hours_end,
+            tzinfo=pytz.UTC,
+        )
+
+        if (day.start < work_start and day.end < work_start) or (day.start > work_end and day.end > work_end):
+            continue
+        elif work_start <= day.start <= work_end and work_start <= day.end <= work_end:
+            result.append((day.start, day.end))
+        elif day.start < work_start and day.end > work_end:
+            result.append((work_start, work_end))
+        elif day.start >= work_start:
+            result.append((day.start, work_end))
+        else:
+            result.append((work_start, day.end))
+
+    return result
+
+
+def get_inside_working_hours_score(events: list[dict], users: dict[str, User]):
+    inside_working_hours_duration = datetime.timedelta(seconds=0)
+
+    for event in events:
+        user_pks = [user["pk"] for user in event["users"]]
+
+        for user_pk in user_pks:
+            start, end = event["start"], event["end"]
+            if event["all_day"]:
+                start = get_day_start(start)
```

Comment:
If I understand correctly, here it makes sense to calculate `start` and `end` for `all_day` events with respect to calendar timezone, because `all_day` event for different timezones will have different `start` and `end` time